pkg/remediator/watch/manager.go

```diff
@@ -327,3 +340,23 @@ func (m *Manager) stopWatcher(gvk schema.GroupVersionKind) {
 	w.Stop()
 	delete(m.watcherMap, gvk)
 }
+
+func (m *Manager) startWatchingCRD(gvk schema.GroupVersionKind, commit string) {
+	m.crdController.SetReconciler(gvk.GroupKind(), func(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) error {
+		gvkMap := map[schema.GroupVersionKind]struct{}{
+			gvk: {},
+		}
+		if customresource.IsEstablished(crd) {
+			return m.AddWatches(ctx, gvkMap, commit)
```

Comment:
When the CRD controller detects a change for a specific CRD, instead of calling`AddWatches` to iterate through all gvks, should we just start the watch for that gvk if it hasn't been started yet?

Reply:
Occurs to me that the Mapper probably needs to be reset though. And it looks to me like the mapper used by the remediator isn't the same as the one used by the reconciler. It gets constructed from DynamicListerWatcherFactoryFromConfig.

I don't actually know if the new discovery API handles evicting stale resource records from the mapper client or if we need to reset it manually, like we do for the reconciler mapper.

Reply:
Actually, it looks like the NewDynamicRESTMapper used by the remediator...
1. Doesn't have a Reset method
2. Automatically calls discoveryClient.ServerResourcesForGroupVersion when a NoMatchError is received, updating to include newly established resources automatically.
3. Lazily discovers resources on-demand

However, there are two caveats:
1. There's no removal of resources from the cache. So RESTMapping won't error after a CRD is deleted.
2. It doesn't use the new aggregated discovery API,  because it was only added upstream 2 weeks ago: https://github.com/kubernetes-sigs/controller-runtime/pull/2901.

So...
- We don't need to reset the mapper when a CRD is newly established.
- There's no obvious way to reset the mapper when a CRD becomes unestablished or deleted. So instead of skipping the next watch start after recreation, the watch start may fail after the RESTMapper succeeds, causing unnecessary errors in the CRD Controller and triggering retry with backoff, until the CRD is recreated (maybe never).

Reply:
I see. The gvkMap is reset to only have one GVK in the crdHandler.

> There's no obvious way to reset the mapper when a CRD becomes unestablished or deleted. So instead of skipping the next watch start after recreation, the watch start may fail after the RESTMapper succeeds, causing unnecessary errors in the CRD Controller and triggering retry with backoff, until the CRD is recreated (maybe never).

I don't quite follow this part. When a CRD becomes unestablished or deleted, won't the handler skips calling `m.AddWatches`? If so, there won't be any errors in the CRD controller. 

Per `RESTMapping won't error after a CRD is deleted`, it sounds like started watch will continue to work without errors.

Reply:
> I don't quite follow this part. When a CRD becomes unestablished or deleted, won't the handler skips calling m.AddWatches? If so, there won't be any errors in the CRD controller.

You're right. The delete/unestablished event doesn't do anything. It used to, while I was prototyping, but I had to disable it because it was causing CR delete events to be missed, because the watch was being stopped prematurely. And it won't error on an established event, because even though the RESTMapper wasn't reset, the resources is back now, so the watch should succeed.

The problem is really just if UpdateWatches or AddWatches is called by the Updater/Applier after a CRD has been deleted. The AddWatches/UpdateWatches will check the RESTMapper, which thinks the resource still exists, and then the startWatch will fail because it actually doesn't. Then that will trigger the applier run loop retry with backoff.

So we need to somehow reset the RESTMapper otherwise the Updater might fail before it has a chance to reapply a managed CRD that was deleted by a third party.

Reply:
If the deleted CRD is managed by RSync, the remediator should be able to re-create the CRD between Applier retries. That will eventually fix the watch start error when the CRD is established.

Alternatively, we can GET the CRD before checking the RESTMapper. If the CRD is unestablished or deleted, we can skip starting the watcher.
This GET call can be omitted if the CRD controller stores the CRD's status in memory.

Reply:
> If the deleted CRD is managed by RSync, the remediator should be able to re-create the CRD between Applier retries. That will eventually fix the watch start error when the CRD is established.

The main edge case this PR is trying to solve is missed delete events for newly created CRs before the Applier succeeds. Eventually starting the watch isn't good enough, because if the CR has already been deleted before the watch starts, the remediator won't know to recreate it. Worst case, if the applier succeeded, but the CR was deleted by a 3rd party before UpdateWatches is called, the CR wouldn't be recreated until the next hourly re-sync.

Plus, the remediator is paused between Applier retries. It only unpauses after the Applier succeeds, because the remediator doesn't follow dependency ordering and could break things worse by trying to remediate after an unsuccessful apply.

> Alternatively, we can GET the CRD before checking the RESTMapper. If the CRD is unestablished or deleted, we can skip starting the watcher. This GET call can be omitted if the CRD controller stores the CRD's status in memory.

The CRDController already gets the CRD from the cache used by the controller-manager informer. And the watch.Manager already starts watches if the CRD is established. The gotcha is that if the RESTMapper cache is not invalidated after a CRD is deleted, then it may contain invalid mappings. Then when the CRD is recreated, the mappings may have changed and the RESTMapper won't know, because it didn't get a NoMatchFound so it didn't refresh from the server. So it needs to auto-invalidate for auto-discovery to be reliable.

So I've added a new ReplaceOnResetRESTMapper that can be used to replace the RESTMapper when Reset is called. Then I added some code to the watch.Manager to handle calling Reset on the mapper when a CRD is established or unestablished, if the mapper still knows about the deleted resource or doesn't know about a new resource.